presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.metadata.BoundVariables;
+import com.facebook.presto.metadata.FunctionKind;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.metadata.SqlScalarFunction;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Primitives;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+import static java.lang.Math.min;
+
+public final class ArrayReduceTwoInputsFunction
+        extends SqlScalarFunction
+{
+    public static final ArrayReduceTwoInputsFunction ARRAY_REDUCE_TWO_INPUTS_FUNCTION = new ArrayReduceTwoInputsFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayReduceTwoInputsFunction.class, "reduce", Type.class, Type.class, Block.class, Block.class, Object.class, MethodHandle.class, MethodHandle.class);
+
+    private ArrayReduceTwoInputsFunction()
+    {
+        super(new Signature(
+                "reduce",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T"), typeVariable("Q"), typeVariable("S"), typeVariable("R")),
```

Comment:
Compared to `T` and `Q`, `T` and `U` is a more common choice. You can also do `T1` and `T2` if that's what you prefer.